.github/actions/publish/action.yml

```
@@ -0,0 +1,87 @@
+---
+name: publish binaries
+description: package GitHub release for node-pre-gyp
+# https://github.com/product-os/flowzone/tree/master/.github/actions
+inputs:
+  json:
+    description: 'JSON stringified object containing all the inputs from the calling workflow'
+    required: true
+  secrets:
+    description: 'JSON stringified object containing all the secrets from the calling workflow'
+    required: true
+  variables:
+    description: 'JSON stringified object containing all the variables from the calling workflow'
+    required: true
+  # --- custom environment
+  NODE_VERSION:
+    description: 'Node.js version'
+    default: '18.x'
+
+runs:
+  # https://docs.github.com/en/actions/creating-actions/creating-a-composite-action
+  using: 'composite'
+  steps:
+    - name: Setup Node.js
+      uses: actions/setup-node@v3
+      with:
+        node-version: ${{ inputs.NODE_VERSION }}
+        registry-url: 'https://registry.npmjs.org'
+
+    - name: Build JSON array from pipe separated list
+      id: config
+      uses: kanga333/json-array-builder@c7cd9d3a8b17cd368e9c2210bc3c16b0e2714ce5 # v0.2.1
+      with:
+        cmd: bash -c "echo $INPUT | tr -d '[:space:]'"
+        separator: "|"
+      env:
+        INPUT: ${{ env.matrix_value }}
+
+    - name: Login to GitHub Container Registry
+      continue-on-error: true
+      uses: docker/login-action@343f7c4344506bcbf9b4de18042ae17996df046d # v3.0.0
+      with:
+        registry: ghcr.io
+        username: ${{ github.actor }}
+        password: ${{ fromJSON(inputs.secrets).GITHUB_TOKEN }}
+
+    - name: Copy the binary from the built docker image
+      shell: bash
+      run: |
+        set -ex
+
+        image="ghcr.io/balena-io-modules/node-systemd:${{ github.event.pull_request.head.sha }}"
+        while ! docker buildx imagetools inspect "${image}" >/dev/null 2>&1
+        do
+          echo "waiting for ${image}..."
+          sleep $(((RANDOM % 10) + 10))
+        done
+
+        # Copy the binary from the image to the local folder for packaging
+        container="node-systemd-$ARCH"
+        docker pull --platform "$PLATFORM" "${image}"
+        docker create --platform "$PLATFORM" --name "${container}" "${image}"
```

```suggestion
        docker run --rm -it -v "$(pwd):/wd" "${image}" cp /usr/src/app/native /wd/native
```
Have you tried this to save some steps?

You also should not need the `--platform` arg when creating the container since you specified it on the image pull.

---

Yes I tried that first but you get exec errors when trying to execute `cp`. I'll remove the `--platform` from docker create

---

The exec errors must be from running containers that are not x86. I would add a step to register QEMU and try that cp command. It will likely be much faster.
https://github.com/product-os/flowzone/blob/e16888d207530b97a017c8fdfcf6f6eecd62670d/.github/workflows/flowzone.yml#L1867-L1872

What you have also works, I just want to verify we know why cp is failing.

---

Yeah, I'm pretty sure the failure was due to the mismatched architecture. [Here](https://github.com/balena-io-modules/node-systemd/actions/runs/6326390666/job/17181135700) is the action run. I didn't know about that QEMU feature. Would it really be faster? The image would still need to be pulled via the `run`, plus the overhead of setting up and booting with qemu. I can still try it as I agree it would end up in cleaner code

---

@klutchell I suppose I can remove the container/image cleanup as I assume those should happen anyway right?

---

Yeah, there shouldn't be any need to do the cleanup yourself.